src/bandersnatch/mirror.py

```diff
@@ -193,6 +109,139 @@ def _filter_packages(self) -> None:
                 else:
                     del self.packages_to_sync[package_name]
 
+    async def determine_packages_to_sync(self) -> None:
+        """
+        Update the self.packages_to_sync to contain packages that need to be
+        synced.
+        """
+        raise NotImplementedError()
+
+    async def package_syncer(self, idx: int) -> None:
+        logger.debug(f"Package syncer {idx} started for duty")
+        while True:
+            try:
+                package = self.package_queue.get_nowait()
+                await package.update_metadata(self.master, attempts=3)
+                await self.process_package(package)
+            except asyncio.QueueEmpty:
+                logger.debug(f"Package syncer {idx} emptied queue")
+                break
+            except PackageNotFound:
+                continue
+            except Exception as e:
+                self.on_error(e, package=package)
+
+    async def process_package(self, package: Package) -> None:
+        raise NotImplementedError()
+
+    async def sync_packages(self) -> None:
+        try:
+            self.package_queue: asyncio.Queue = asyncio.Queue()
+            # Sorting the packages alphabetically makes it more predictable:
+            # easier to debug and easier to follow in the logs.
+            for name in sorted(self.packages_to_sync):
+                serial = int(self.packages_to_sync[name])
+                await self.package_queue.put(Package(name, serial=serial))
+
+            sync_coros: List[Awaitable] = [
+                self.package_syncer(idx) for idx in range(self.workers)
+            ]
+            try:
+                await asyncio.gather(*sync_coros)
+            except KeyboardInterrupt as e:
+                self.on_error(e)
+        except (ValueError, TypeError) as e:
+            # This is for when self.packages_to_sync isn't of type Dict[str, int]
+            self.on_error(e)
+
+    def finalize_sync(self) -> None:
+        raise NotImplementedError()
+
+    def on_error(self, exception: BaseException, **kwargs: Dict) -> None:
+        raise NotImplementedError()
+
+
+class BandersnatchMirror(Mirror):
+
+    need_index_sync = True
+    json_save = False  # Whether or not to mirror PyPI JSON metadata to disk
```

Comment:
A lot of these class variables are overwritten by ```__init__```.   Let's double check the defaults below to make sure they match up, move the comments, and then delete the superfluous definitions here.